In [189]:
import pandas as pd

df = pd.read_csv("C:\\Users\In9.var\Desktop\Projects\data\diabetes.csv")
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [190]:
print(df.shape)

(768, 9)


In [191]:
print(df[df['Outcome'] == 1].shape)
print(df[df['Outcome'] == 0].shape)
print(df[df['Outcome'] == 1].shape[0] / df[df['Outcome'] == 0].shape[0])
print(df.isna().sum())
y= df['Outcome']
X= df.drop(["Outcome"], axis=1)

(268, 9)
(500, 9)
0.536
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


In [192]:
X

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

In [193]:
y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

In [194]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler

scaler = StandardScaler()
X_st = scaler.fit_transform(X)
X_st

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])

In [195]:
scaler = MinMaxScaler()
X_norm = scaler.fit_transform(X)
X_norm

array([[0.35294118, 0.74371859, 0.59016393, ..., 0.50074516, 0.23441503,
        0.48333333],
       [0.05882353, 0.42713568, 0.54098361, ..., 0.39642325, 0.11656704,
        0.16666667],
       [0.47058824, 0.91959799, 0.52459016, ..., 0.34724292, 0.25362938,
        0.18333333],
       ...,
       [0.29411765, 0.6080402 , 0.59016393, ..., 0.390462  , 0.07130658,
        0.15      ],
       [0.05882353, 0.63316583, 0.49180328, ..., 0.4485842 , 0.11571307,
        0.43333333],
       [0.05882353, 0.46733668, 0.57377049, ..., 0.45305514, 0.10119556,
        0.03333333]])

In [196]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)

In [197]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix

scores = []
for i in range (1, 15):
    knn = KNeighborsClassifier(i)
    
    knn.fit(X_train, y_train)
    
    scores.append(knn.score(X_test, y_test))
print(scores)

[0.62109375, 0.67578125, 0.703125, 0.69921875, 0.6875, 0.69921875, 0.7265625, 0.69921875, 0.703125, 0.703125, 0.71484375, 0.71484375, 0.703125, 0.71484375]


In [198]:
max_score = max(scores)
scores_ind = [i for i, v in enumerate(scores) if v == max_score]
scores_ind

[6]

In [199]:
X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
scores = []
for i in range (1, 15):
    knn = KNeighborsClassifier(i)
    
    knn.fit(X_train, y_train)
    
    scores.append(knn.score(X_test, y_test))
print(scores)
max_score = max(scores)
scores_ind = [i for i, v in enumerate(scores) if v == max_score]
scores_ind

[0.68359375, 0.6796875, 0.6796875, 0.69140625, 0.6796875, 0.69140625, 0.7109375, 0.7109375, 0.71484375, 0.6953125, 0.6875, 0.69140625, 0.6953125, 0.69921875]


[8]

In [200]:
# соседи станд.
X_train, X_test, y_train, y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)

knn = KNeighborsClassifier(5)

knn.fit(X_train,y_train)

y_pred = knn.predict(X_test)

confusion_matrix(y_test, y_pred)

array([[133,  37],
       [ 43,  43]], dtype=int64)

In [201]:
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score
print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))

Accuracy: 0.6875
Recall: 0.5
Precision_score: 0.5375
f1: 0.5180722891566264


In [202]:
# knieghbors для норм данных

X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)

knn = KNeighborsClassifier(7)

knn.fit(X_train,y_train)

y_pred = knn.predict(X_test)

confusion_matrix(y_test, y_pred)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))
confusion_matrix(y_test, y_pred)
# немного лучше но плохо

Accuracy: 0.7109375
Recall: 0.5465116279069767
Precision_score: 0.573170731707317
f1: 0.5595238095238095


array([[135,  35],
       [ 39,  47]], dtype=int64)

In [203]:
# логистическая регрессия для норм. данных

from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model = model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))
confusion_matrix(y_test, y_pred)
# куда лучше

Accuracy: 0.7421875
Recall: 0.5348837209302325
Precision_score: 0.6388888888888888
f1: 0.5822784810126581


array([[144,  26],
       [ 40,  46]], dtype=int64)

In [204]:
#логистическая регрессия для стард. данных

X_train, X_test, y_train, y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)

model = LogisticRegression()
model = model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))
confusion_matrix(y_test, y_pred)

Accuracy: 0.734375
Recall: 0.6162790697674418
Precision_score: 0.6022727272727273
f1: 0.6091954022988506


array([[135,  35],
       [ 33,  53]], dtype=int64)

Лучший recall = 0.627906976744186 у логистической регрессии для стард. данных
Лучший accuracy = 0.7421875 у логистической регрессии для норм. данных
Лучший precision = 0.63888888888 у логистической регрессии для норм. данных
Но лучшей моделью я считаю последнюю (лог. рег. для  стандарт. данных), тк у неё лучшая F-мера, последняя модель более вариативная в различных задачах с разными целями

In [205]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1/3, random_state=42)
model = LogisticRegression()
model = model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))
confusion_matrix(y_test, y_pred)

Accuracy: 0.75
Recall: 0.627906976744186
Precision_score: 0.627906976744186
f1: 0.627906976744186


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


array([[138,  32],
       [ 32,  54]], dtype=int64)

Оказывается лучшим по всем параметрам, кроме Precision оказалась лог. регр. для базовых данных